libraries/persistent_store/src/fragment.rs

```diff
@@ -0,0 +1,179 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Helper functions for fragmented entries.
+//!
+//! This module permits to handle entries larger than the [maximum value
+//! length](Store::max_value_length) by storing ordered consecutive fragments in a sequence of keys.
+//! The first keys hold fragments of maximal length, followed by a possibly partial fragment. The
+//! remaining keys are not used.
+
+use crate::{Storage, Store, StoreError, StoreHandle, StoreResult, StoreUpdate};
+use alloc::vec::Vec;
+use core::ops::Range;
+
+/// Represents a sequence of keys.
+pub trait Keys {
+    /// Returns the number of keys.
+    fn len(&self) -> usize;
+
+    /// Returns the position of a key in the sequence.
+    fn pos(&self, key: usize) -> Option<usize>;
+
+    /// Returns the key of a position in the sequence.
+    ///
+    /// # Preconditions
+    ///
+    /// The position must be within the length: `pos < len()`.
+    fn key(&self, pos: usize) -> usize;
+}
+
+impl Keys for Range<usize> {
+    fn len(&self) -> usize {
+        self.end - self.start
+    }
+
+    fn pos(&self, key: usize) -> Option<usize> {
+        if self.start <= key && key < self.end {
+            Some(key - self.start)
+        } else {
+            None
+        }
+    }
+
+    fn key(&self, pos: usize) -> usize {
+        debug_assert!(pos < Keys::len(self));
+        self.start + pos
+    }
+}
+
+/// Reads the concatenated value of a sequence of keys.
+pub fn read(store: &Store<impl Storage>, keys: &impl Keys) -> StoreResult<Option<Vec<u8>>> {
+    let handles = get_handles(store, keys)?;
+    if handles.is_empty() {
+        return Ok(None);
+    }
+    let mut result = Vec::with_capacity(handles.len() * store.max_value_length());
+    for handle in handles {
+        result.extend(handle.get_value(store)?);
+    }
+    Ok(Some(result))
+}
+
+/// Reads a range from the concatenated value of a sequence of keys.
+///
+/// This is equivalent to calling [`read`] then taking the range except that:
+/// - Only the needed chunks are read.
+/// - The range is truncated to fit in the value.
+pub fn read_range(
+    store: &Store<impl Storage>,
+    keys: &impl Keys,
+    range: Range<usize>,
+) -> StoreResult<Option<Vec<u8>>> {
+    let range_len = match range.end.checked_sub(range.start) {
+        None => return Err(StoreError::InvalidArgument),
+        Some(x) => x,
+    };
+    let handles = get_handles(store, keys)?;
+    if handles.is_empty() {
+        return Ok(None);
+    }
+    let mut result = Vec::with_capacity(range_len);
+    let mut offset = 0;
+    for handle in handles {
+        let start = range.start.saturating_sub(offset);
+        let length = handle.get_length(store)?;
+        let end = core::cmp::min(range.end.saturating_sub(offset), length);
+        offset += length;
+        if start < end && end <= length {
+            result.extend(&handle.get_value(store)?[start..end]);
+        }
```

Comment:
Do we want to `break` if `offset >= range.end`?

Reply:
I don't think it's useful, but we can. It's just CPU and negligible to the rest. At first I wasn't even sure it's useful to have `read_range` compare to just calling `read` and taking the range after. I'm still not sure actually, but since you had it in your PR, I did it here too.

Reply:
I don't have a strong preference, so resolved.